Storage/SQL/SQLiteHistory.swift

```diff
@@ -915,49 +915,66 @@ extension SQLiteHistory: SyncableHistory {
         // A difficulty here: we don't want to fetch *all* visits, only some number of the most recent.
         // (It's not enough to only get new ones, because the server record should contain more.)
         //
-        // That's the greatest-N-per-group problem in SQL. Please read and understand the solution
-        // to this (particularly how the LEFT OUTER JOIN/HAVING clause works) before changing this query!
+        // That's the greatest-N-per-group problem. We used to do this in SQL, joining
+        // the visits table to a subselect of visits table, however, ran into OOM issues (Bug 1417034)
+        //
+        // Now, we want a more dumb approach with no joins in SQL and doing group-by-site and limit-to-N in swift.
+        //
+        // We do this in a single query, rather than the N+1 that desktop takes.
         //
-        // We can do this in a single query, rather than the N+1 that desktop takes.
         // We then need to flatten the cursor. We do that by collecting
         // places as a side-effect of the factory, producing visits as a result, and merging in memory.
 
         // Turn our lazy collection of integers into a comma-seperated string for the IN clause.
         let historyIDs = Array(places.keys)
-        let inClause = "siteID IN ( \(historyIDs.map(String.init).joined(separator: ",")) )"
 
         let sql =
-        "SELECT v1.siteID AS siteID, v1.date AS visitDate, v1.type AS visitType " +
-        "FROM (" +
-        "   SELECT * FROM \(TableVisits) WHERE \(inClause) AND type <> 0" +
-        ") AS v1 " +
-        "LEFT OUTER JOIN \(TableVisits) AS v2 ON v1.siteID = v2.siteID AND v1.date < v2.date " +
-        "GROUP BY v1.date " +
-        "HAVING COUNT(*) < ? " +
-        "ORDER BY v1.siteID, v1.date DESC"
-
-        // Seed our accumulator with empty lists since we already know which IDs we will be fetching.
-        var visits = [Int: [Visit]]()
-        historyIDs.forEach { visits[$0] = [] }
-
-        // Add each visit to its history item's list.
-        let visitsAccumulator: (SDRow) -> Void = { row in
-            let date = row.getTimestamp("visitDate")!
-            let type = VisitType(rawValue: row["visitType"] as! Int)!
-            let visit = Visit(date: date, type: type)
-            let id = row["siteID"] as! Int
-            visits[id]?.append(visit)
-        }
+                "SELECT siteID, date AS visitDate, type AS visitType " +
+                "FROM \(TableVisits) " +
+                "WHERE siteID IN (\(historyIDs.map(String.init).joined(separator: ","))) " +
+                "ORDER BY date DESC"
```

Comment:
Why not `ORDER BY siteID DESC, date DESC`? Or do you need to do only date in order to get an index walk?

Reply:
You're right:

with `ORDER BY siteID DESC, date DESC`

```
⦿ EXPLAIN QUERY ---------------- 
⦿ EXPLAIN: id: 0, order: 0, from: 0, details: SEARCH TABLE visits USING COVERING INDEX sqlite_autoindex_visits_1 (siteID=?)
⦿ EXPLAIN: id: 0, order: 0, from: 0, details: EXECUTE LIST SUBQUERY 1
```

with `ORDER BY date DESC`: 

```
⦿ EXPLAIN QUERY ---------------- 
⦿ EXPLAIN: id: 0, order: 0, from: 0, details: SEARCH TABLE visits USING COVERING INDEX sqlite_autoindex_visits_1 (siteID=?)
⦿ EXPLAIN: id: 0, order: 0, from: 0, details: EXECUTE LIST SUBQUERY 1
⦿ EXPLAIN: id: 0, order: 0, from: 0, details: USE TEMP B-TREE FOR ORDER BY
```